src/bin/pg_autoctl/service_keeper.c

```diff
@@ -596,6 +500,144 @@ keeper_node_active_loop(Keeper *keeper, pid_t start_pid)
 }
 
 
+/*
+ * keeper_node_active calls the node_active function on the monitor, and when
+ * it could contact the monitor it also updates our copy of the list of other
+ * nodes currenty in the group (keeper->otherNodes).
+ *
+ * keeper_node_active returns true if it could successfully connect to the
+ * monitor, and false otherwise. When it returns false, it also checks for
+ * network partitions and set the goal state to DEMOTE_TIMEOUT_STATE when
+ * needed.
+ */
+static bool
+keeper_node_active(Keeper *keeper)
+{
+	KeeperConfig *config = &(keeper->config);
+	KeeperStateData *keeperState = &(keeper->state);
+	Monitor *monitor = &(keeper->monitor);
+	LocalPostgresServer *postgres = &(keeper->postgres);
+
+	uint64_t now = time(NULL);
+	MonitorAssignedState assignedState = { 0 };
+
+	char expectedSlotName[BUFSIZE] = { 0 };
+
+	bool forceCacheInvalidation = false;
+	bool reportPgIsRunning = ReportPgIsRunning(keeper);
+
+	/* We used to output that in INFO every 5s, which is too much chatter */
+	log_debug("Calling node_active for node %s/%d/%d with current state: "
+			  "%s, "
+			  "PostgreSQL %s running, "
+			  "sync_state is \"%s\", "
+			  "current lsn is \"%s\".",
+			  config->formation,
+			  keeperState->current_node_id,
+			  keeperState->current_group,
+			  NodeStateToString(keeperState->current_role),
+			  reportPgIsRunning ? "is" : "is not",
+			  postgres->pgsrSyncState,
+			  postgres->currentLSN);
+
+
+	/* ensure we use the correct retry policy with the monitor */
+	(void) pgsql_set_main_loop_retry_policy(&(monitor->pgsql));
+
+	/*
+	 * Report the current state to the monitor and get the assigned state.
+	 */
+	if (!monitor_node_active(monitor,
+							 config->formation,
+							 config->nodename,
+							 config->pgSetup.pgport,
+							 keeperState->current_node_id,
+							 keeperState->current_group,
+							 keeperState->current_role,
+							 reportPgIsRunning,
+							 postgres->currentLSN,
+							 postgres->pgsrSyncState,
+							 &assignedState))
+	{
+		log_error("Failed to get the goal state from the monitor");
+
+		/*
+		 * Check whether we're likely to be in a network partition.
+		 * That will cause the assigned_role to become demoted.
+		 */
+		if (keeperState->current_role == PRIMARY_STATE)
+		{
+			log_warn("Checking for network partitions...");
+
+			if (!is_network_healthy(keeper))
+			{
+				keeperState->assigned_role = DEMOTE_TIMEOUT_STATE;
```

Comment:
NOTE: This behaviour described in this comment also applies the code on the `master` branch. However, it seems wrong or at least very confusing there also.

After reading and understanding this code better I realised that this assigned role will never actually become the `current_role`. The reason for this is that `needsStateChange` is only set when `couldContactMonitor` is `true`. Which is always `false` since this code is only executed when the monitor could not be reached. This means postgres is never actually turned off when `DEMOTE_TIMEOUT_STATE` is set.

I tried this out locally:
1. Start monitor
2. Start primary
3. Start secondary
4. kill both monitor and secondary
5. Wait 20 seconds

Expected behaviour:
6. the primary postgres is turned off 
Actual behaviour:
6. we start getting these log messages:
```
16:11:59 31917 WARN  Checking for network partitions...
16:11:59 31917 INFO  Failed to contact the monitor or standby in 18446744073709551581 seconds, at 20 seconds we shut down PostgreSQL to prevent split brain issues
16:11:59 31917 INFO  Network in not healthy, switching to state demote_timeout
```
However `psql -p {primaryPort}` still works. Write queries are blocked though, since the secondary cannot be reached and it is in `sync_stanbys`. 

I think the main problem that this behaviour causes is that clients will still connect to this primary, since `target_session_attrs=read-write` is still true for this server. 

Side-issue, the seconds in the log message are clearly incorrect:
```
Failed to contact the monitor or standby in 18446744073709551581 seconds
```

Reply:
It looks to me that the culprit is a change that (IIRC) comes with this PR, where we protected FSM changes when we failed to contact the monitor, which is the wrong thing to do as you're showing. So to me, on a first reading, it boils down to the following change that I have just added:

``` diff
modified   src/bin/pg_autoctl/service_keeper.c
@@ -372,13 +372,21 @@ keeper_node_active_loop(Keeper *keeper, pid_t start_pid)
 
 		couldContactMonitor = couldContactMonitorThisRound;
 
-		if (couldContactMonitor &&
-			keeperState->assigned_role != keeperState->current_role)
+		if (keeperState->assigned_role != keeperState->current_role)
 		{
 			needStateChange = true;
 
```

Please have a look at this change.

Reply:
I see that it's also broken on master in a different way now, so the breakage has been introduced earlier. It still looks to me that the current approach should fix it in this PR.